[2022.10.21] feat: 토스트 컴포넌트 구현

rktnsinger opened the pull request:
## 작업 사항
- 간단한 알림을 보낼 수 있는 토스트 메시지 컴포넌트입니다.
- props로 text와 color를 전달 가능합니다
- useToast 훅을 이용하여 재사용 가능합니다
- 애니메이션 기본 지속시간은 4초입니다

https://user-images.githubusercontent.com/96766719/197126724-d07ad517-7d70-4f8e-ac55-bfda9b3bbb05.mov


## 잘 봐주세요
- 컴포넌트 재사용이 용이할지 생각해봐주세요.

## PR 전 확인사항
- [x] 가장 최신 브랜치를 pull했습니다.
- [x] base 브랜치명을 확인했습니다.
- [x] 코드 컨벤션을 모두 지켰습니다.
- [x] 적절한 라벨이 있습니다.
- [x] assignee가 있습니다.

## 비고
- 모달과 토스트 메시지에서 Portal 컴포넌트를 재사용할 수 있도록, id를 props로 전달받아 별개의 div 태그를 생성하는 Portal 컴포넌트를 추가하였습니다.

Commit:
feat: 토스트 컴포넌트 구현
- 간단한 알림을 띄우는 토스트 컴포넌트 제작
- 모달창과 토스트컴포넌트에서 Portal을 재사용하기 위하여 별도의 Portal 컴포넌트 제작
- useToast 커스텀 훅으로 토스트 메시지 재사용

Commit:
style: 컴포넌트 내 엘리먼트 이름 변경

Yanguk commented on src/components/common/Portal/index.js:
```diff
@@ -0,0 +1,45 @@
+  const [parentElement, removeParentElement] = createParentElement(id);
+
+  useEffect(() => {
+    return () => removeParentElement();
```
언마운트되면 포탈 element가 사라지네요 재밌는 코드 같습니다 👍

Yanguk commented on src/components/common/ToastMessage/index.js:
```diff
@@ -0,0 +1,91 @@
+  padding: 10px 20px;
+  border: none;
+  border-radius: 10px;
+  background-color: ${(props) => props.color};
```
스타일드 컴포넌트의 활용 👍

Yanguk reviewed (approved):
재사용성이 용이하게 포탈과 토스트컴프넌트를 훌륭 하게 만들어 주신 것 같아요 고생하셨습니다! 👍

ssong-yi commented on src/components/common/ToastMessage/index.js:
```diff
@@ -0,0 +1,91 @@
+import Portal from "../Portal";
+import theme from "../../../config/constants/theme";
+
+const DEFAULT_TIME_OUT = 4000;
```
이 상수는 다른 파일로 빼서 관리하지 않아도 될까요?

rktnsinger replied:
반영하겠습니다! 혹시 파일이름 뭐가 좋을까요..
그냥 이거 하나만 넣을거면 defaultTimeOut 이럼되는데
혹시 나중에 추가될지 모르니 defaultValues 이렇게할까요?

ssong-yi replied:
백엔드처럼 constants 폴더 내 index.js 파일 내에 넣는 넣는 거 어떠세요?

rktnsinger replied:
좋아요!! 반영하겠습니다 ㅎㅎ

ssong-yi commented on src/components/common/ToastMessage/index.js:
```diff
@@ -0,0 +1,91 @@
+      opacity: 0;
+    }
+  }
+`;
```
이렇게 구현해주셨군요! 배워갑니다~

ssong-yi commented on src/components/common/ToastMessage/index.js:
```diff
@@ -0,0 +1,91 @@
+  setSendToast: PropTypes.func.isRequired,
+};
+
+ToastMessage.defaultProps = {
+  color: theme.gray,
+};
```
`defaultProps`로 정해주셔서, 컴포넌트를 이용하기 편하면서도, 다양한 옵션을 지정할 수 있게 해주셨네요! 👍

ssong-yi commented on src/hooks/useToast.js:
```diff
@@ -0,0 +1,22 @@
+export default function useToast(initialState = false) {
+  const [sendToast, setSendToast] = useState(initialState);
+
+  const handleSendToast = () => setSendToast(true);
```
미리 정한 컨벤션에 따라, 한 줄이지만 아래처럼 해주는 게 어떨까요?
```suggestion
  const handleSendToast = () => {
    return setSendToast(true)
  };
```

rktnsinger replied:
앗 이부분 지난번에 양욱님과 한번 이야기 나눈적 있는데 80자 안넘어서 한줄로 쓰긴했습니다! 모달과 컨벤션을 맞추려구요 ㅎㅎ 괜찮을까요??

혹 바꿔야하면 수정될 부분들이 좀 더 늘어날 것 같긴 합니다..!

ssong-yi replied:
아네넵! 간결하게 쓰는 것도 좋죠! 저도 고려하면서 작업하겠습니다~

ssong-yi commented on src/hooks/useToast.js:
```diff
@@ -0,0 +1,22 @@
+
+import ToastMessage from "../components/common/ToastMessage";
+
+export default function useToast(initialState = false) {
+  const [sendToast, setSendToast] = useState(initialState);
```
default 파라미터를 잘 사용해주셨네요! 👍

ssong-yi reviewed (commented):
형우님 리뷰 확인부탁드립니다~

Commit:
refactor: 코드리뷰 반영 및 기타 변경사항 반영
- 토스트 메시지 지속기간 상수화하여 별도 파일에 선언
- 버전관리: 0.0.2 -> 0.0.3
- lint rule 추가: "no-underscore-dangle"

rktnsinger commented:
리뷰 반영하였습니다! 확인 부탁드려요~~

ssong-yi reviewed (approved):
고생하셨습니다~
